akka-stream-tests/src/test/scala/akka/stream/io/OutputStreamSourceSpec.scala

```diff
@@ -0,0 +1,135 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.io
+
+import java.io.IOException
+import java.util.concurrent.TimeoutException
+
+import akka.stream._
+import akka.stream.scaladsl.Keep
+import akka.stream.testkit.Utils._
+import akka.stream.testkit._
+import akka.stream.testkit.scaladsl.TestSink
+import akka.util.ByteString
+
+import scala.concurrent.duration.Duration.Zero
+import scala.concurrent.duration._
+import scala.concurrent.{ Await, Future }
+
+class OutputStreamSourceSpec extends AkkaSpec(UnboundedMailboxConfig) {
+  import system.dispatcher
+
+  val settings = ActorMaterializerSettings(system)
+  implicit val materializer = ActorMaterializer(settings)
+
+  val timeout = 300.milliseconds
+  val bytes = "abc".getBytes
+  val byteString = ByteString("abc")
+
+  def expectBlocked[T](f: Future[T]) =
+    the[Exception] thrownBy Await.result(f, timeout) shouldBe a[TimeoutException]
+
+  def expectSuccess[T](f: Future[T], value: T) =
+    Await.result(f, timeout) should be(value)
+
+  "OutputStreamSource" must {
+    "read bytes from OutputStream" in assertAllStagesStopped {
+      val (outputStream, probe) = OutputStreamSource().toMat(TestSink.probe[ByteString])(Keep.both).run
+      val s = probe.expectSubscription()
+
+      outputStream.write(bytes)
+      s.request(1)
+      probe.expectNext(byteString)
+      outputStream.close()
+      probe.expectComplete()
+      //Await.result(f, timeout) should be(3)
+    }
+
+    "block flush call until send all buffer to downstream" in assertAllStagesStopped {
+      val (outputStream, probe) = OutputStreamSource().toMat(TestSink.probe[ByteString])(Keep.both).run
+      val s = probe.expectSubscription()
+
+      outputStream.write(bytes)
+      val f = Future(outputStream.flush())
+
+      expectBlocked(f)
+      probe.expectNoMsg(Zero)
+
+      s.request(1)
+      expectSuccess(f, ())
+      probe.expectNext(byteString)
+
+      outputStream.close()
+      probe.expectComplete()
+    }
+
+    "not block flushes when buffer is empty" in assertAllStagesStopped {
+      val (outputStream, probe) = OutputStreamSource().toMat(TestSink.probe[ByteString])(Keep.both).run
+      val s = probe.expectSubscription()
+
+      outputStream.write(bytes)
+
+      val f = Future(outputStream.flush())
+      s.request(1)
+      expectSuccess(f, ())
+      probe.expectNext(byteString)
+
+      val f2 = Future(outputStream.flush())
+      expectSuccess(f2, ())
+
+      outputStream.close()
+      probe.expectComplete()
+    }
+
+    "block writes when buffer is full" in assertAllStagesStopped {
+      val (outputStream, probe) = OutputStreamSource().toMat(TestSink.probe[ByteString])(Keep.both).run
+      val s = probe.expectSubscription()
+
+      //default max-input-buffer-size = 16
```

Comment:
Ok, will it be like GraphStage.withAttributes fro set and then GraphStage.module.attributes to read?